### PR TITLE
[FLINK-4535] [cluster management] resourceManager process the registration from TaskExecutor

### PR DESCRIPTION
**Pull request by beyond1920**
This pull request is to implement ResourceManager registration with TaskExecutor, which including:
1. Check whether input resourceManagerLeaderId is as same as the current leadershipSessionId of resourceManager. If not, it means that maybe two or more resourceManager exists at the same time, and current resourceManager is not the proper rm. so it rejects or ignores the registration.
2. Check whether exists a valid taskExecutor at the giving address by connecting to the address. Reject the registration from invalid address. (which is hidden in the connect method)
3. Keep resourceID and taskExecutorGateway mapping relationships, And optionally keep resourceID and container mapping relationships in yarn mode.
4. Send registration successful ack to the taskExecutor.

Main difference are 3 points:
1. Add UnmatchedLeaderSessionIDException to specify that received leader session ID is not as same as expected.
2. Change registerTaskExecutor method  of ResourceManager
3. Add a test class for ResourceManager

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/rpc/UnmatchedLeaderSessionIDException.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.rpc;
+
+import java.util.UUID;
+
+/**
+ * An exception specifying that received leader session ID is not as same as expected.
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rpc/UnmatchedLeaderSessionIDException.java]: "that the received", "is not the same as"

File: flink-runtime/src/main/java/org/apache/flink/runtime/rpc/UnmatchedLeaderSessionIDException.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ */
+public class UnmatchedLeaderSessionIDException extends Exception {
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rpc/UnmatchedLeaderSessionIDException.java]: I think it's ok to name this exception `LeaderSessionIDException`

File: flink-runtime/src/main/java/org/apache/flink/runtime/rpc/UnmatchedLeaderSessionIDException.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+	/** expected leader session id */
+	private final UUID expectedLeaderSessionID;
+
+	/** actual leader session id */
+	private final UUID actualLeaderSessionID;
+
+	public UnmatchedLeaderSessionIDException(UUID expectedLeaderSessionID, UUID actualLeaderSessionID) {
+		super("Unmatched leader session ID : expected " + expectedLeaderSessionID + ", actual " + actualLeaderSessionID);
+		this.expectedLeaderSessionID = expectedLeaderSessionID;
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rpc/UnmatchedLeaderSessionIDException.java]: `checkNotNull`

File: flink-runtime/src/main/java/org/apache/flink/runtime/rpc/UnmatchedLeaderSessionIDException.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+		this.actualLeaderSessionID = actualLeaderSessionID;
+	}
+
+	/**
+	 * Get expect leader session id
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rpc/UnmatchedLeaderSessionIDException.java]: "expected"

File: flink-runtime/src/main/java/org/apache/flink/runtime/rpc/resourcemanager/ResourceManager.java [hunk 5]
```diff
@@ -138,22 +150,46 @@ public SlotAssignment requestSlot(SlotRequest slotRequest) {
 
 
 	/**
-	 *
-	 * @param resourceManagerLeaderId  The fencing token for the ResourceManager leader 
+	 * Register a taskExecutor at the resource manager
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rpc/resourcemanager/ResourceManager.java]: Mixed camel and non-camel case. Furthermore, the `taskExecutor` is not the class name. Line break missing.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rpc/resourcemanager/ResourceManager.java [hunk 6]
```diff
@@ -138,22 +150,46 @@ public SlotAssignment requestSlot(SlotRequest slotRequest) {
...
+			public RegistrationResponse apply(final TaskExecutorGateway taskExecutorGateway) {
+				InstanceID instanceID = null;
+				if(startedTaskExecutorGateways.containsKey(resourceID)) {
+					log.warn("Receive a duplicate registration from TaskExecutor {} at ({})", resourceID, taskExecutorAddress);
+					instanceID = taskExecutorGateways.get(startedTaskExecutorGateways.get(resourceID));
+				} else {
+					startedTaskExecutorGateways.put(resourceID, taskExecutorGateway);
+					instanceID = new InstanceID();
+					taskExecutorGateways.put(taskExecutorGateway, instanceID);
+				}
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rpc/resourcemanager/ResourceManager.java]: line break missing

File: flink-runtime/src/main/java/org/apache/flink/runtime/rpc/resourcemanager/ResourceManager.java [hunk 7]
```diff
@@ -138,22 +150,46 @@ public SlotAssignment requestSlot(SlotRequest slotRequest) {
...
-			ResourceID resourceID) {
+	public Future<RegistrationResponse> registerTaskExecutor(
+		final UUID resourceManagerLeaderId,
+		final String taskExecutorAddress,
+		final ResourceID resourceID) {
+
+		if(!leaderSessionID.equals(resourceManagerLeaderId)) {
+			log.warn("Discard registration from TaskExecutor {} at ({}) because the expected leader session ID {} did not equal the received leader session ID  {}",
+				resourceID, taskExecutorAddress, leaderSessionID, resourceManagerLeaderId);
+			return Futures.failed(new UnmatchedLeaderSessionIDException(leaderSessionID, resourceManagerLeaderId));
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rpc/resourcemanager/ResourceManager.java]: Should we fail or decline the registration here? So either sending an exception or a `RegistrationResponse.Decline` message.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rpc/resourcemanager/ResourceManager.java [hunk 8]
```diff
@@ -53,14 +56,23 @@
  */
 public class ResourceManager extends RpcEndpoint<ResourceManagerGateway> {
 	private final Map<JobMasterGateway, InstanceID> jobMasterGateways;
+
+	/** ResourceID and TaskExecutorGateway mapping relationship of registered taskExecutors */
+	private final Map<ResourceID, TaskExecutorGateway>  startedTaskExecutorGateways;
+
+	/** TaskExecutorGateway and InstanceId mapping relationship of registered taskExecutors */
+	private final Map<TaskExecutorGateway, InstanceID> taskExecutorGateways;
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rpc/resourcemanager/ResourceManager.java]: Wouldn't it make sense to group the `TaskExecutorGateway` and the `InstanceID` into a `TaskExecutorRegistration` class which is stored under the resource ID? Then we would get rid of a lookup when accessing the `InstanceID` given the resource ID.

File: flink-runtime/src/test/java/org/apache/flink/runtime/rpc/resourcemanager/ResourceManagerTest.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+		resourceManager.start();
+		final UUID leaderSessionId = UUID.randomUUID();
+		leaderElectionService.isLeader(leaderSessionId);
+
+		// test throw exception when receive a registration from taskExecutor which takes unmatched leaderSessionId
+		UUID differentLeaderSessionID = UUID.randomUUID();
+		Future<RegistrationResponse> unMatchedLeaderFuture = resourceManager.registerTaskExecutor(differentLeaderSessionID, taskExecutorAddress, taskExecutorResourceID);
+		assertTrue(unMatchedLeaderFuture.isCompleted());
+		assertTrue(unMatchedLeaderFuture.failed().isCompleted());
+		assertTrue(unMatchedLeaderFuture.failed().value().get().get() instanceof UnmatchedLeaderSessionIDException);
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rpc/resourcemanager/ResourceManagerTest.java]: Why not simply doing `Await.result` and catching the `UnmatchedLeaderSessionIDException`?

**Comment by tillrohrmann** [comment]
Thanks for the contribution @beyond1920. I've had some minor inline comments. After addressing them, it should be good to merge.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rpc/resourcemanager/ResourceManager.java (hunk 7 above)
Comment (beyond1920) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rpc/resourcemanager/ResourceManager.java]: Do you prefer to send a Decline message under the condition?

**Comment by beyond1920** [comment]
@tillrohrmann , thanks for your review. I changed the pr based on your advice, including ：
1. Modify UnmatchedLeaderSessionIDException to LeaderSessionIDException, format code style and method comment, and add checkNotNull to constructor
2. Format code style and method comment of ResourceManager
3. Group the TaskExecutorGateway and the InstanceID into a TaskExecutorRegistration
4. Modify testcases which are expected to throw exception at ResourceManagerTest

Besides, I don't understand this comment of yours: "Should we fail or decline the registration here? So either sending an exception or a RegistrationResponse.Decline message.". Do you means you prefer to sending a Decline message?

**Comment by mxm** [comment]
Thank you for the pull request! I've rebased the changes and will merge them to the master.

**Comment by mxm** [comment]
This has been merged to `flip-6`. Could you please close the PR?

**Pull request closed without merging**